Index: FFmpeg/libavutil/pixdesc.c
===================================================================
--- FFmpeg.orig/libavutil/pixdesc.c
+++ FFmpeg/libavutil/pixdesc.c
@@ -3267,6 +3267,30 @@ static const AVPixFmtDescriptor av_pix_f
         .name = "ohcodec",
         .flags = AV_PIX_FMT_FLAG_HWACCEL,
     },
+    [AV_PIX_FMT_NV15] = {
+        .name = "nv15",
+        .nb_components = 3,
+        .log2_chroma_w = 1,
+        .log2_chroma_h = 1,
+        .comp = {
+            { 0, 10,  0, 0, 10 },      /* Y */
+            { 1, 20,  0, 0, 10 },      /* U */
+            { 1, 20, 10, 0, 10 },      /* V */
+        },
+        .flags = AV_PIX_FMT_FLAG_PLANAR | AV_PIX_FMT_FLAG_BITSTREAM,
+    },
+    [AV_PIX_FMT_NV20] = {
+        .name = "nv20",
+        .nb_components = 3,
+        .log2_chroma_w = 1,
+        .log2_chroma_h = 0,
+        .comp = {
+            { 0, 10,  0, 0, 10 },      /* Y */
+            { 1, 20,  0, 0, 10 },      /* U */
+            { 1, 20, 10, 0, 10 },      /* V */
+        },
+        .flags = AV_PIX_FMT_FLAG_PLANAR | AV_PIX_FMT_FLAG_BITSTREAM,
+    },
 };
 
 static const char * const color_range_names[] = {
Index: FFmpeg/libavutil/pixfmt.h
===================================================================
--- FFmpeg.orig/libavutil/pixfmt.h
+++ FFmpeg/libavutil/pixfmt.h
@@ -196,8 +196,8 @@ enum AVPixelFormat {
     AV_PIX_FMT_XYZ12LE,      ///< packed XYZ 4:4:4, 36 bpp, (msb) 12X, 12Y, 12Z (lsb), the 2-byte value for each X/Y/Z is stored as little-endian, the 4 lower bits are set to 0
     AV_PIX_FMT_XYZ12BE,      ///< packed XYZ 4:4:4, 36 bpp, (msb) 12X, 12Y, 12Z (lsb), the 2-byte value for each X/Y/Z is stored as big-endian, the 4 lower bits are set to 0
     AV_PIX_FMT_NV16,         ///< interleaved chroma YUV 4:2:2, 16bpp, (1 Cr & Cb sample per 2x1 Y samples)
-    AV_PIX_FMT_NV20LE,       ///< interleaved chroma YUV 4:2:2, 20bpp, (1 Cr & Cb sample per 2x1 Y samples), little-endian
-    AV_PIX_FMT_NV20BE,       ///< interleaved chroma YUV 4:2:2, 20bpp, (1 Cr & Cb sample per 2x1 Y samples), big-endian
+    AV_PIX_FMT_NV20LE,       ///< interleaved chroma YUV 4:2:2, 20bpp, (1 Cr & Cb sample per 2x1 Y samples), little-endian, deprecated in favor of AV_PIX_FMT_NV20
+    AV_PIX_FMT_NV20BE,       ///< interleaved chroma YUV 4:2:2, 20bpp, (1 Cr & Cb sample per 2x1 Y samples), big-endian, deprecated in favor of AV_PIX_FMT_NV20
 
     AV_PIX_FMT_RGBA64BE,     ///< packed RGBA 16:16:16:16, 64bpp, 16R, 16G, 16B, 16A, the 2-byte value for each R/G/B/A component is stored as big-endian
     AV_PIX_FMT_RGBA64LE,     ///< packed RGBA 16:16:16:16, 64bpp, 16R, 16G, 16B, 16A, the 2-byte value for each R/G/B/A component is stored as little-endian
@@ -499,6 +499,9 @@ enum AVPixelFormat {
 
     AV_PIX_FMT_OHCODEC, /// hardware decoding through openharmony
 
+    AV_PIX_FMT_NV15,        ///< like P010, but has no zero padding bits, 15bpp, bitstream
+    AV_PIX_FMT_NV20,        ///< like P210, but has no zero padding bits, 20bpp, bitstream
+
     AV_PIX_FMT_NB         ///< number of pixel formats, DO NOT USE THIS if you want to link with shared libav* because the number of formats might differ between versions
 };
 
@@ -597,7 +600,6 @@ enum AVPixelFormat {
 #define AV_PIX_FMT_YUVA444P16 AV_PIX_FMT_NE(YUVA444P16BE, YUVA444P16LE)
 
 #define AV_PIX_FMT_XYZ12      AV_PIX_FMT_NE(XYZ12BE, XYZ12LE)
-#define AV_PIX_FMT_NV20       AV_PIX_FMT_NE(NV20BE,  NV20LE)
 #define AV_PIX_FMT_AYUV64     AV_PIX_FMT_NE(AYUV64BE, AYUV64LE)
 #define AV_PIX_FMT_P010       AV_PIX_FMT_NE(P010BE,  P010LE)
 #define AV_PIX_FMT_P012       AV_PIX_FMT_NE(P012BE,  P012LE)
Index: FFmpeg/libswscale/format.c
===================================================================
--- FFmpeg.orig/libswscale/format.c
+++ FFmpeg/libswscale/format.c
@@ -228,6 +228,8 @@ static const LegacyFormatEntry legacy_fo
     [AV_PIX_FMT_XYZ12LE]        = { 1, 1, 1 },
     [AV_PIX_FMT_AYUV64LE]       = { 1, 1},
     [AV_PIX_FMT_AYUV64BE]       = { 1, 1 },
+    [AV_PIX_FMT_NV15]           = { 1, 0 },
+    [AV_PIX_FMT_NV20]           = { 1, 0 },
     [AV_PIX_FMT_P010LE]         = { 1, 1 },
     [AV_PIX_FMT_P010BE]         = { 1, 1 },
     [AV_PIX_FMT_P012LE]         = { 1, 1 },
Index: FFmpeg/libswscale/input.c
===================================================================
--- FFmpeg.orig/libswscale/input.c
+++ FFmpeg/libswscale/input.c
@@ -946,6 +946,39 @@ static void nv21ToUV_c(uint8_t *dstU, ui
     nvXXtoUV_c(dstV, dstU, src1, width);
 }
 
+static av_always_inline void nv15_20ToYUV_c(uint16_t *dst, const uint8_t *src,
+                                            int dst_pos, int src_pos)
+{
+    int shift = (src_pos << 1) & 7;
+    src_pos = (src_pos * 10) >> 3;
+    AV_WN16(dst + dst_pos,
+            ((AV_RL16(src + src_pos) >> shift) |
+             (AV_RL16(src + src_pos + 1) << (8 - shift))) & 0x3FF);
+}
+
+static void nv15_20ToY_c(uint8_t *_dst, const uint8_t *_src, const uint8_t *unused1,
+                         const uint8_t *unused2, int width, uint32_t *unused, void *opq)
+{
+    int i;
+    const uint8_t *src = (const uint8_t *)_src;
+    uint16_t *dst      = (uint16_t *)_dst;
+    for (i = 0; i < width; i++)
+        nv15_20ToYUV_c(dst, src, i, i);
+}
+
+static void nv15_20ToUV_c(uint8_t *_dstU, uint8_t *_dstV,
+                          const uint8_t *unused0, const uint8_t *_src1, const uint8_t *_src2,
+                          int width, uint32_t *unused, void *opq)
+{
+    int i;
+    const uint8_t *src1 = (const uint8_t *)_src1;
+    uint16_t *dstU      = (uint16_t *)_dstU, *dstV = (uint16_t *)_dstV;
+    for (i = 0; i < width; i++) {
+        nv15_20ToYUV_c(dstU, src1, i, 2 * i);
+        nv15_20ToYUV_c(dstV, src1, i, 2 * i + 1);
+    }
+}
+
 #define p01x_uv_wrapper(fmt, shift) \
     static void fmt ## LEToUV ## _c(uint8_t *dstU,                       \
                                        uint8_t *dstV,                    \
@@ -2038,6 +2071,10 @@ av_cold void ff_sws_init_input_funcs(Sws
     case AV_PIX_FMT_XV48BE:
         *chrToYV12 = read_xv48be_UV_c;
         break;
+    case AV_PIX_FMT_NV15:
+    case AV_PIX_FMT_NV20:
+        *chrToYV12 = nv15_20ToUV_c;
+        break;
     case AV_PIX_FMT_NV20LE:
         *chrToYV12 = nv20LEToUV_c;
         break;
@@ -2626,6 +2663,10 @@ av_cold void ff_sws_init_input_funcs(Sws
     case AV_PIX_FMT_BGRA64LE:
         *lumToYV12 = bgr64LEToY_c;
         break;
+    case AV_PIX_FMT_NV15:
+    case AV_PIX_FMT_NV20:
+        *lumToYV12 = nv15_20ToY_c;
+        break;
     case AV_PIX_FMT_NV20LE:
         *lumToYV12 = nv20LEToY_c;
         break;
Index: FFmpeg/libswscale/swscale_unscaled.c
===================================================================
--- FFmpeg.orig/libswscale/swscale_unscaled.c
+++ FFmpeg/libswscale/swscale_unscaled.c
@@ -269,6 +269,62 @@ static int nv24ToYuv420Wrapper(SwsIntern
     return srcSliceH;
 }
 
+static int nv15_20ToPlanarWrapper(SwsInternal *c, const uint8_t *const src8[],
+                                  const int srcStride[], int srcSliceY,
+                                  int srcSliceH, uint8_t *const dstParam[],
+                                  const int dstStride[])
+{
+    const AVPixFmtDescriptor *src_format = av_pix_fmt_desc_get(c->opts.src_format);
+    const AVPixFmtDescriptor *dst_format = av_pix_fmt_desc_get(c->opts.dst_format);
+    int vsub = 1 << dst_format->log2_chroma_h;
+    const uint8_t **src = (const uint8_t**)src8;
+    uint16_t *dstY = (uint16_t*)(dstParam[0] + dstStride[0] * srcSliceY);
+    uint16_t *dstU = (uint16_t*)(dstParam[1] + dstStride[1] * srcSliceY / vsub);
+    uint16_t *dstV = (uint16_t*)(dstParam[2] + dstStride[2] * srcSliceY / vsub);
+    int x, y;
+
+    /* Calculate net shift required for values. */
+    const int shift[3] = {
+        dst_format->comp[0].depth + dst_format->comp[0].shift -
+        src_format->comp[0].depth - src_format->comp[0].shift,
+        dst_format->comp[1].depth + dst_format->comp[1].shift -
+        src_format->comp[1].depth - src_format->comp[1].shift,
+        dst_format->comp[2].depth + dst_format->comp[2].shift -
+        src_format->comp[2].depth - src_format->comp[2].shift,
+    };
+
+    for (y = srcSliceH; y > 0; y--) {
+        const uint8_t *tsrcY = src[0];
+        uint16_t *tdstY = dstY;
+        for (x = c->opts.src_w / 4; x > 0; x--) {
+            *tdstY++ = (((tsrcY[1] & 0x3 ) << 8) | (tsrcY[0]        & 0xFF)) << shift[0];
+            *tdstY++ = (((tsrcY[2] & 0xF ) << 6) | ((tsrcY[1] >> 2) & 0x3F)) << shift[0];
+            *tdstY++ = (((tsrcY[3] & 0x3F) << 4) | ((tsrcY[2] >> 4) & 0xF )) << shift[0];
+            *tdstY++ = (((tsrcY[4] & 0xFF) << 2) | ((tsrcY[3] >> 6) & 0x3 )) << shift[0];
+            tsrcY += 5;
+        }
+        src[0] += srcStride[0];
+        dstY += dstStride[0] / sizeof(uint16_t);
+    }
+
+    for (y = srcSliceH / vsub; y > 0; y--) {
+        const uint8_t *tsrcUV = src[1];
+        uint16_t *tdstU = dstU, *tdstV = dstV;
+        for (x = c->chrSrcW / 2; x > 0; x--) {
+            *tdstU++ = (((tsrcUV[1] & 0x3 ) << 8) | (tsrcUV[0]        & 0xFF)) << shift[1];
+            *tdstV++ = (((tsrcUV[2] & 0xF ) << 6) | ((tsrcUV[1] >> 2) & 0x3F)) << shift[2];
+            *tdstU++ = (((tsrcUV[3] & 0x3F) << 4) | ((tsrcUV[2] >> 4) & 0xF )) << shift[1];
+            *tdstV++ = (((tsrcUV[4] & 0xFF) << 2) | ((tsrcUV[3] >> 6) & 0x3 )) << shift[2];
+            tsrcUV += 5;
+        }
+        src[1] += srcStride[1];
+        dstU += dstStride[1] / sizeof(uint16_t);
+        dstV += dstStride[2] / sizeof(uint16_t);
+    }
+
+    return srcSliceH;
+}
+
 static int planarToP01xWrapper(SwsInternal *c, const uint8_t *const src8[],
                                const int srcStride[], int srcSliceY,
                                int srcSliceH, uint8_t *const dstParam8[],
@@ -2411,6 +2467,19 @@ void ff_get_unscaled_swscale(SwsInternal
         (srcFormat == AV_PIX_FMT_NV24 || srcFormat == AV_PIX_FMT_NV42)) {
         c->convert_unscaled = nv24ToPlanarWrapper;
     }
+    /* nv15_to_yuv420p1x & nv20_to_yuv422p1x */
+    if ((srcFormat == AV_PIX_FMT_NV15 &&
+         (dstFormat == AV_PIX_FMT_YUV420P10 ||
+          dstFormat == AV_PIX_FMT_YUV420P12 ||
+          dstFormat == AV_PIX_FMT_YUV420P14 ||
+          dstFormat == AV_PIX_FMT_YUV420P16)) ||
+        (srcFormat == AV_PIX_FMT_NV20 &&
+         (dstFormat == AV_PIX_FMT_YUV422P10 ||
+          dstFormat == AV_PIX_FMT_YUV422P12 ||
+          dstFormat == AV_PIX_FMT_YUV422P14 ||
+          dstFormat == AV_PIX_FMT_YUV422P16))) {
+        c->convert_unscaled = nv15_20ToPlanarWrapper;
+    }
     /* yuv2bgr */
     if ((srcFormat == AV_PIX_FMT_YUV420P || srcFormat == AV_PIX_FMT_YUV422P ||
          srcFormat == AV_PIX_FMT_YUVA420P) && isAnyRGB(dstFormat) &&
